File: server/internal/httpapi/textgolden_test.go
package httpapi

import (
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"

	"stencil/server/internal/testutil"
)

// Byte-exact goldens for the server's pinned LLM text surface: the two system-prompt
// heads and the GET /llm/info body. Both are slated to move into JSON assets, so these
// pin the current bytes to prove the move is verbatim. Rewrite with
// SERVER_UPDATE_GOLDENS=1 go test ./internal/httpapi/...

// goldenPath resolves tests/goldens beside this file, the way llmprompt_test.go finds
// the canonical asset.
func goldenPath(name string) string {
	_, thisFile, _, ok := runtime.Caller(0)
	if !ok {
		panic("runtime.Caller failed")
	}
	return filepath.Join(filepath.Dir(thisFile), "goldens", name)
}

func checkGolden(t *testing.T, name, actual string) {
	t.Helper()
	path := goldenPath(name)
	if os.Getenv("SERVER_UPDATE_GOLDENS") == "1" {
		if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
			t.Fatalf("create the goldens dir: %v", err)
		}
		if err := os.WriteFile(path, []byte(actual), 0o644); err != nil {
			t.Fatalf("write %s: %v", path, err)
		}
		return
	}
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v — rerun with SERVER_UPDATE_GOLDENS=1", path, err)
	}
	if string(raw) == actual {
		return
	}
	t.Errorf("golden %s differs (SERVER_UPDATE_GOLDENS=1 to rewrite):\n%s",
		name, goldenDiff(string(raw), actual))
}

// goldenDiff reports the first few differing lines, so a failure names the wording
// that moved rather than just saying "not equal".
func goldenDiff(expected, actual string) string {
	want := strings.Split(expected, "\n")
	got := strings.Split(actual, "\n")
	n := len(want)
	if len(got) > n {
		n = len(got)
	}
	var b strings.Builder
	shown := 0
	for i := 0; i < n && shown < 12; i++ {
		w, g := "<eof>", "<eof>"
		if i < len(want) {
			w = want[i]
		}
		if i < len(got) {
			g = got[i]
		}
		if w == g {
			continue
		}
		fmt.Fprintf(&b, "  line %d:\n    want %s\n    got  %s\n", i+1, w, g)
		shown++
	}
	if shown == 0 {
		fmt.Fprintf(&b, "  identical line-wise; lengths %d vs %d\n", len(expected), len(actual))
	}
	return b.String()
}

func TestPromptHeadGoldens(t *testing.T) {
	checkGolden(t, "llm_editor_prompt_head.txt", llmEditorPromptHead)
	checkGolden(t, "llm_extension_prompt_head.txt", llmExtensionPromptHead)
}

func TestLLMInfoBodyGolden(t *testing.T) {
	var b strings.Builder
	for _, tc := range []struct {
		name string
		llm  LLM
	}{
		{"disabled", nil},
		{"enabled (model claude-test)", &testutil.EchoLLM{}},
	} {
		api := llmAPI(t, tc.llm)
		tok := issueToken(t, api, "")
		rec := do(t, api, http.MethodGet, "/llm/info", tok, nil)
		fmt.Fprintf(&b, "== GET /llm/info — %s\n%d %s\n%s\n",
			tc.name, rec.Code, rec.Header().Get("Content-Type"), rec.Body.String())
	}
	checkGolden(t, "llm_info.txt", b.String())
}

File: mcp/tests/text_golden_test.rs
//! Byte-exact goldens for the MCP server's user-facing prose: every `#[tool]` description
//! and the `ServerHandler::get_info()` instructions. That text is slated to
//! move into JSON assets, so these pin the current bytes to prove the move is verbatim.
//! Rewrite with `MCP_UPDATE_GOLDENS=1 cargo test`.

use rmcp::ServerHandler;
use stencil_mcp::server::StencilServer;

const GOLDENS_DIR: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/tests/goldens");

/// Compare against `tests/goldens/{name}`, or rewrite it when `MCP_UPDATE_GOLDENS=1`.
fn check(name: &str, actual: &str) {
    let path = format!("{GOLDENS_DIR}/{name}");
    if std::env::var("MCP_UPDATE_GOLDENS").as_deref() == Ok("1") {
        std::fs::create_dir_all(GOLDENS_DIR).expect("create the goldens dir");
        std::fs::write(&path, actual).unwrap_or_else(|e| panic!("cannot write {path}: {e}"));
        return;
    }
    let expected = std::fs::read_to_string(&path).unwrap_or_else(|e| {
        panic!("cannot read {path}: {e} — rerun with MCP_UPDATE_GOLDENS=1")
    });
    if expected == actual {
        return;
    }
    panic!(
        "golden {name} differs (MCP_UPDATE_GOLDENS=1 to rewrite):\n{}",
        diff(&expected, actual)
    );
}

/// The first few differing lines, so a failure names the wording that moved.
fn diff(expected: &str, actual: &str) -> String {
    let want: Vec<&str> = expected.lines().collect();
    let got: Vec<&str> = actual.lines().collect();
    let mut out = String::new();
    let mut shown = 0;
    for i in 0..want.len().max(got.len()) {
        let (w, g) = (want.get(i).unwrap_or(&"<eof>"), got.get(i).unwrap_or(&"<eof>"));
        if w == g || shown >= 12 {
            continue;
        }
        out.push_str(&format!("  line {}:\n    want {w}\n    got  {g}\n", i + 1));
        shown += 1;
    }
    if shown == 0 {
        out.push_str(&format!(
            "  identical line-wise; lengths {} vs {}\n",
            expected.len(),
            actual.len()
        ));
    }
    out
}

/// The description the `#[tool]` attribute registered, as an MCP client sees it.
fn description(tool: rmcp::model::Tool) -> String {
    tool.description
        .unwrap_or_else(|| panic!("tool \"{}\" has no description", tool.name))
        .to_string()
}

#[test]
fn the_tool_descriptions_match_their_goldens() {
    check(
        "stencil_edit.txt",
        &description(StencilServer::stencil_edit_tool_attr()),
    );
    check(
        "stencil_probe.txt",
        &description(StencilServer::stencil_probe_tool_attr()),
    );
    check(
        "stencil_prompt.txt",
        &description(StencilServer::stencil_prompt_tool_attr()),
    );
    check(
        "stencil_script.txt",
        &description(StencilServer::stencil_script_tool_attr()),
    );
    check(
        "source_site.txt",
        &description(StencilServer::source_site_tool_attr()),
    );
}

#[test]
fn the_server_instructions_match_their_golden() {
    let info = StencilServer::default().get_info();
    check(
        "instructions.txt",
        &info.instructions.expect("get_info() carries instructions"),
    );
}
